compiler/parser/Parser.y

```diff
@@ -1273,10 +1274,39 @@ flex_choice_decl :: { Located FlexChoiceData }
                        , cdChoiceDoc = $6 }
     }
 
-observer_and_controller :: { (LHsExpr GhcPs, Maybe (LHsExpr GhcPs)) }
-  : 'controller' party_list { (applyConcat $2, Nothing) }
-  -- we currently only support an optional observer clause *before* the controller clause
-  | 'observer' parties 'controller' party_list  { (applyConcat $4, Just (applyConcat $2)) }
+choice_parties :: { Located [Located ChoiceBodyDecl] }
+  : legacy_observer_and_controller      { $1 }
+  | 'where' choice_body_decl_list       { $2 }
+
+choice_body_decl_list :: { Located [Located ChoiceBodyDecl] }
+  : '{' choice_body_decls '}'           { $2 }
+  | vocurly choice_body_decls close     { $2 }
+
+-- new choice_body_decls syntax: follows a 'where' keyword
+-- (optional) 'observer', 'authority' and (mandatory) 'controller' decls can be in any order
+
+choice_body_decls :: { Located [Located ChoiceBodyDecl] }
+  : choice_body_decls ';' choice_body_decl      { sLL $1 $> $ $3 : (unLoc $1) }
+  | choice_body_decls ';'                       { sLL $1 $> $ unLoc $1 }
+  | choice_body_decl                            { sL1 $1 [$1] }
+  | {- empty -}                                 { sL0 [] }
+
+choice_body_decl :: { Located ChoiceBodyDecl }
+  : controller_decl                             { sL1 $1 $ ChoiceControllerDecl $1 }
+  | observer_decl                               { sL1 $1 $ ChoiceObserverDecl $1 }
+  | authority_decl                              { sL1 $1 $ ChoiceAuthorityDecl $1 }
+
+-- legacy choice_body_decls syntax; without a 'where' keyword
+-- supports just an optional observer clause *before* a mandatory controller clause
```

Comment:
Is there any reason to lock this at that order? Is the intent to entirely phase out the legacy syntax, or support both?

Reply:
This was the only order which was supported by the legacy syntax.
The new syntax mandates a `where` keyword, as this allows the more flexible ordering.

Reply:
Phasing out is a worthy goal. But to be honest we still support the legacy controller syntax, and that became legacy more than 3 years ago when we introduced flexible choices. So don't hold your breath!

Reply:
Right! If we're semi-expecting this legacy syntax to still be supported in years to come, should we not consider supporting any order?
Or can we properly deprecate it ASAP?

Reply:
Ah I see, its a technical issue, okay I'm satisfied :)